src/tasks/decode.ts

```diff
@@ -0,0 +1,333 @@
+import "hardhat-deploy";
+import "@nomiclabs/hardhat-ethers";
+
+import { promises as fs } from "fs";
+
+import chalk from "chalk";
+import { BigNumber, BigNumberish, Contract, utils, constants } from "ethers";
+import { task } from "hardhat/config";
+
+import {
+  decodeTradeFlags,
+  EncodedSettlement,
+  Trade,
+  BUY_ETH_ADDRESS,
+  Interaction,
+  SigningScheme,
+} from "../ts";
+
+const WIDTH = 120;
+const INVALID_TOKEN = " ! Invalid token ! ";
+const NATIVE_TOKEN = " native token ";
+
+interface Token {
+  contract: Contract;
+  symbol: string | null;
+  decimals: number | null;
+  address: string;
+  nativeFlag: boolean;
+  price: BigNumber | undefined;
+  index: number;
+}
+
+type MaybeToken =
+  | Token
+  | {
+      index: number;
+      decimals?: undefined;
+      symbol?: undefined;
+      nativeFlag?: undefined;
+    };
+
+type FormatToken = {
+  address: string;
+  index: string;
+  symbol: string;
+  price: string;
+};
+
+function formatToken(token: Token): FormatToken {
+  return {
+    address: token.address,
+    index: token.index.toString(),
+    symbol: token.nativeFlag ? NATIVE_TOKEN : token.symbol ?? INVALID_TOKEN,
+    price: (token.price ?? "no price").toString(),
+  };
+}
+
+function cellText(text: string, size: number): string {
+  const inner =
+    text.length > size - 2
+      ? text.slice(0, size - 5) + "..."
+      : text.padStart(size - 2, " ");
+  return " " + inner + " ";
+}
+
+const mainLabel = (s: string) => chalk.bold(chalk.yellow(s + ":"));
+const label = (s: string) => chalk.cyan(s + ":");
+
+function displayTokensColumnWidth(
+  formattedTokens: FormatToken[],
+): Record<keyof FormatToken, number> {
+  const width = {
+    address: 42,
+    index: 5,
+    symbol: 20,
+    price: 40,
+  };
+
+  const maxWidth = (key: keyof typeof width) =>
+    formattedTokens.reduce((max, token) => Math.max(max, token[key].length), 0);
+  for (const key in width) {
+    const typedKey = key as keyof typeof width;
+    // pad with a space left and right (+2)
+    width[typedKey] =
+      Math.max(Math.min(width[typedKey], maxWidth(typedKey)), typedKey.length) +
+      2;
+  }
+  return width;
+}
+
+function displayTokens(tokens: Token[]) {
+  const headers = ["address", "index", "symbol", "price"] as const;
+  const formattedTokens = tokens.map(formatToken);
+  const columnWidth = displayTokensColumnWidth(formattedTokens);
+  console.log(chalk.bold("=== Decoded tokens ==="));
```

Comment:
nit: Maybe just 
```suggestion
  console.log(chalk.bold("=== Tokens ==="));
```

Since `Decoded` is added to all sections and doesn't add too much value in my opinion (seeing as the task is already called `decode`).